packages/rest-crud/src/crud-rest.controller.ts

```diff
@@ -0,0 +1,320 @@
+//
+// First of all, to correctly decorate methods and define schemas for request
+// and response bodies, we need to know the target model which will be used by
+// the controller. As a result, this file has to export a function that will
+// create a constructor class specific to the given model.
+//
+// Secondly, TypeScript does not allow decorators to be used in class
+// expressions - see https://github.com/microsoft/TypeScript/issues/7342.
+// As a result, we cannot write implement the factory as `return class ...`,
+// but have to define the class as an internal type and return the controller
+// constructor in a new statement.
+// Because the controller class is an internal type scoped to the body of the
+// factory function, we cannot use it to describe the return type. We must
+// explicitly provide the return type.
+//
+// To work around those issues, we use the following design:
+// - The interface `CrudRestController` describes controller methods (members)
+// - The type `CrudRestControllerCtor` describes the class constructor.
+// - `defineCrudRestController` returns `CrudRestControllerCtor` type.
+
+/**
+ * This interface describes prototype members of the controller class
+ * returned by `defineCrudRestController`.
+ */
+export interface CrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+> {
+  /**
+   * The backing repository used to access & modify model data.
+   */
+  readonly repository: EntityCrudRepository<T, IdType>;
+
+  /**
+   * Implementation of the endpoint `POST /`.
+   * @param data Model data
+   */
+  create(data: Omit<T, IdName>): Promise<T>;
+}
+
+/**
+ * Constructor of the controller class returned by `defineCrudRestController`.
+ */
+export interface CrudRestControllerCtor<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+> {
+  new (
+    repository: EntityCrudRepository<T, IdType, Relations>,
+  ): CrudRestController<T, IdType, IdName, Relations>;
+}
+
+/**
+ * Options to configure different aspects of a CRUD REST Controller.
+ */
+export interface CrudRestControllerOptions {
+  /**
+   * The base path where to "mount" the controller.
+   */
+  basePath: string;
+}
+
+/**
+ * Create (define) a CRUD Controller class for the given model.
+ *
+ * Example usage:
+ *
+ * ```ts
+ * const CrudRestController = defineCrudRestController<
+ *   Product,
+ *   typeof Product.prototype.id,
+ *   'id'
+ * >(Product, {basePath: '/products'});
+ *
+ * class ProductController extends CrudRestController {
+ *   constructor() {
+ *    super(repo);
+ *   }
+ * }
+ *
+ * app.controller(ProductController);
+ * ```
+ *
+ * @param modelCtor A model class, e.g. `Product`.
+ * @param options Configuration options, e.g. `{basePath: '/products'}`.
+ */
+export function defineCrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+>(
```

Comment:
Is it feasible to define a `CrudRestController` from a `Repository` (instance or class) directly? Ideally, if a developer adds a repository using `lb4 repository` to connect a model to a data source, we should allows him/her to expose the repository as CRUD REST APIs out of box.

Reply:
That's an interesting idea, we can explore it later if you like. This new module is in 0.x version, so even if we need to make breaking changes to support your idea, then that's ok.

Our current plan is to allow developers to go from `lb4 model` to REST API with no custom repository class needed - see https://github.com/strongloop/loopback-next/issues/2036

The idea is that the user will tell us that let's say `Product` model should be exposed via CRUD REST API (as opposed to e.g. KeyValue REST API) using datasource `db`, and the framework will take care of the rest.

Reply:
The current implementation requires a repository class to be defined. I added a few comments inline. I'm open to incremental improvements.